### ISSUE TITLE
Shopping cart product options missing

### ISSUE THREAD

**Issue by mengento** [magento/magento2#5612]
## Steps to reproduce
1. Install Magento 2.1
2. Add Customizable Options . 
3. Add  Product to Shopping cart.
4. Not change any attributes(Customizable Options ), save the product again,
5. Refresh Shopping cart ,
## Expected result
1. Shopping car product options normal
## Actual result
1. Shopping car product options missing
2. Product page Options Id Change

![20160713172524](https://cloud.githubusercontent.com/assets/10205547/16798716/073f6224-4920-11e6-87aa-a49b9343bb6c.png)

**Comment by veloraven** [comment]
@mengento thank you for your report.
Internal ticket MAGETWO-56410 was created for this issue.

**Comment by erikhansen** [comment]
@veloraven This is a pretty serious issue and is affecting at least one of our M2 merchants. Can you provide an update on an estimated fix for this issue?

**Comment by erikhansen** [comment]
@veloraven Do you have any update for me on this?

**Comment by manfrinm** [comment]
@veloraven news about this?

**Comment by manfrinm** [comment]
Hello, I've created this patch [CustomOptions Patch](https://github.com/Hevelop/magento2-patches/blob/master/Patch-Magento_Catalog-M2.1.5-MAGETWO-56410-MAGETWO-56411-github-issues-5931-5612.patch) that solve this issue and the Custom option ID issue [#5931](https://github.com/magento/magento2/issues/5931)
The patch can be applied on CE 2.1.5 or 2.1.4 and on EE 2.1.5 or 2.1.4 but I've tested this patch only for Magento EE 2.1.5.
Hopes is helpful

**Comment by kaushalsutharhs** [comment]
@manfrinm  : I'm getting error, Using CE 2.1.5

```
 Applying patches for magento/module-catalog
    https://raw.githubusercontent.com/Hevelop/magento2-patches/master/Patch-Magento_Catalog-M2.1.5-MAGETWO-56410-MAGETWO-56411-github-issues-5931-5612.patch (Fix: https://github.com/magento/magento2/issues/5612)
   Could not apply patch! Skipping.
```

**Comment by manfrinm** [comment]
@kaushalsutharhs I've try to apply the patch on Magento CE 2.1.4  and 2.1.5 base version, and it goes correctly. Try to execute with -vvv to get more information about it.
Anyway, I've talked with official magento's developers (EE support), and they told me that they are currently on development with the "official patch"

**Comment by kaushalsutharhs** [comment]
Hi @manfrinm , Thanks for the reply, I'm getting following error while applying patch.

```
Executing command (CWD): cd '/home/test/Projects/magento2project/vendor/magento/module-catalog' && GIT_DIR=. git apply --check '-p1' '/tmp/58c2415a1bd56.patch'
error: patch failed: Ui/DataProvider/Product/Form/Modifier/CustomOptions.php:112
error: Ui/DataProvider/Product/Form/Modifier/CustomOptions.php: patch does not apply

cd '/home/test/Projects/magento2project/vendor/magento/module-catalog' && GIT_DIR=. git apply --check '-p0' '/tmp/58c2415a1bd56.patch'
Executing command (CWD): cd '/home/test/Projects/magento2project/vendor/magento/module-catalog' && GIT_DIR=. git apply --check '-p0' '/tmp/58c2415a1bd56.patch'
error: b/Controller/Adminhtml/Product/Initialization/Helper.php: No such file or directory
error: b/Model/Product/Option/Repository.php: No such file or directory
error: b/Model/Product/Option/SaveHandler.php: No such file or directory
error: b/Model/Product/Option/Validator/Select.php: No such file or directory
error: b/Model/Product/Option/Value.php: No such file or directory
error: b/Model/ResourceModel/Product/Option.php: No such file or directory
error: b/Model/ResourceModel/Product/Option/Value.php: No such file or directory
error: b/Ui/DataProvider/Product/Form/Modifier/CustomOptions.php: No such file or directory

cd '/home/test/Projects/magento2project/vendor/magento/module-catalog' && GIT_DIR=. git apply --check '-p2' '/tmp/58c2415a1bd56.patch'
Executing command (CWD): cd '/home/test/Projects/magento2project/vendor/magento/module-catalog' && GIT_DIR=. git apply --check '-p2' '/tmp/58c2415a1bd56.patch'
error: Adminhtml/Product/Initialization/Helper.php: No such file or directory
error: Product/Option/Repository.php: No such file or directory
error: Product/Option/SaveHandler.php: No such file or directory
error: Product/Option/Validator/Select.php: No such file or directory
error: Product/Option/Value.php: No such file or directory
error: ResourceModel/Product/Option.php: No such file or directory
error: ResourceModel/Product/Option/Value.php: No such file or directory
error: DataProvider/Product/Form/Modifier/CustomOptions.php: No such file or directory


```

**Comment by sheldmandu** [comment]
We are also experiencing this issue. In addition to shopping cart options it also affects Reorder functionality.  This is also a problem not just for custom options on simple products but also for bundle options, which should probably be raised as a separate issue.

**Comment by scottsb** [comment]
@kaushalsutharhs: I had the same issue applying the patch. I have created [an updated patch](https://github.com/ripenecommerce/magento2-patches/blob/master/Patch-Magento_Catalog-M2.1.5-MAGETWO-56410-MAGETWO-56411-github-issues-5931-5612.patch) to fix the issue with `CustomOptions.php` and have further removed a number of other changes that @manfrinm made that are good but unnecessary for the actual fix (such as improving docblocks or clarifying variable names). This will hopefully make the patch more robust if Magento releases additional versions before fixing this bug.

**Comment by veloraven** [comment]
This issue is fixed in internal ticket MAGETWO-57153.
It will be available in one of the nearest releases.

**Comment by scottsb** [comment]
@veloraven, I couldn't find any commits mentioning that internal ticket number. Is this going to be released for 2.1.x or 2.2?

**Comment by LiamKarlMitchell** [comment]
What version of Magento 2 was this fixed in officially?
